operational_analysis/methods/quality_check_automation.py

```diff
@@ -1,67 +1,181 @@
+        self._t_utc = f"{time_field}_utc"
+        self._t_local = f"{time_field}_localized"
         self._id = id_field
         self._freq = freq
         self._lat_lon = lat_lon
-        self._dst_subset = dst_subset
-        self._check_tz = check_tz
+        self._local_tz = local_tz
+        self._local_ptz = pytz.timezone(local_tz)
+        self._tz_aware = timezone_aware
+        self._offset = "utc_offset"
+        self._dst = "is_dst"
+        self._non_dst_offset = self._local_ptz.localize(datetime(2021, 1, 1)).utcoffset()
 
         if self._id is None:
             self._id = "ID"
             self._df["ID"] = "Data"
 
+        self._convert_datetime_column()
+
+    def _determine_offset_dst(self, df: pd.DataFrame) -> None:
+        """Creates a column of "utc_offset" and "is_dst".
+
+        Args:
+         df(:obj:`pd.DataFrame`): The dataframe object to manipulate.
+
+        Returns:
+         (:obj:`pd.DataFrame`): The updated dataframe with "utc_offset" and "is_dst" columns created.
+        """
+        dt = df.copy().tz_convert(self._local_tz)
+        dt_col = dt.index.to_pydatetime()
+
+        # Determine the Daylight Savings Time status and UTC offset
+        dt[self._offset] = [el.utcoffset() for el in dt_col]
+        dt[self._dst] = (dt[self._offset] != self._non_dst_offset).astype(bool)
+
+        # Convert back to UTC
+        dt = dt.tz_convert("UTC")
+        return dt
+
+    def _convert_datetime_column(self) -> None:
+        """Converts the passed timestamp data to a pandas-encoded Datetime, and creates a
+        corresponding localized and UTC timestamp using the `time_field` column name with either
+        "localized" or "utc", respectively.
+        """
+        # Convert the timestamps to datetime.datetime objects
+        dt_col = self._df[self._t].values
+
+        # Check for raw timestamp inputs or pre-formatted
+        if isinstance(dt_col[0], str):
+            dt_col = [dateutil.parser.parse(el) for el in dt_col]
+
+        if self._tz_aware:
+            pd_dt_col = pd.to_datetime(dt_col, utc=True).tz_convert(self._local_tz)
+            self._df[self._t_local] = pd_dt_col
+        else:
+            pd_dt_col = pd.to_datetime(dt_col)
+            self._df[self._t_local] = pd_dt_col.tz_localize(self._local_tz, ambiguous=True)
+
+        self._df[self._t] = pd_dt_col
+        self._df = self._df.set_index(pd.DatetimeIndex(self._df[self._t_local]))
+
+        # Create the UTC-converted time-stamp
+        try:
+            self._df[self._t_utc] = pd.to_datetime(
+                [el.astimezone(tz.tzutc()) for el in dt_col]
+            ).tz_convert("UTC")
+        except AttributeError:  # catches numpy datetime error for astimezone() not existing
+            self._df = self._df.tz_convert("UTC")
+            self._df[self._t_utc] = self._df.index
+
+        self._df = self._determine_offset_dst(self._df)
+
+    @abstractmethod
     @logged_method_call
     def run(self):
```

Comment:
That makes sense, especially given recent v3 conversations on whether or not we should have users implement the functionality or us provide a way to use it more easily. I assume this would be all the same processes would run except the WTK features?

Reply:
yeah, exactly. Thanks for adding this.